### ISSUE TITLE
BASIS textures assets not usable with create cubemap

### ISSUE THREAD

**Issue by gandhars** [playcanvas/engine#1983]
When creating a CUBEMAP asset from BASIS compressed textures, the cubemap not created. If I use JPG files, everything works fine.
I don't need create prefiltered cubemap, I need use cubemap with custom shader as 'samplerCube'  uniform on posteffect.

https://forum.playcanvas.com/t/basis-texture-for-cubemap/12505

**Comment by slimbuck** [comment]
#1983! :)

**Comment by gandhars** [comment]
Hello Donovan,
thank you for informing me. Is possible to view to patch before it will be
officially released ?
 Radek Hart

st 13. 5. 2020 v 12:22 odesílatel Donovan Hutchence <
notifications@github.com> napsal:

**Comment by slimbuck** [comment]
Sorry @gandhars... I am busy working on cubemaps and confused while going through the tickets confused ticket 1983 with 1984 (the book :) ). I'll let you know when the fix is ready (though it probably won't be very soon).

**Comment by willeastcott** [comment]
Reopening because #2108 broke the After The Flood project, so I backed it out. @slimbuck - please have another go at a fix. 😄

**Comment by leonidaspir** [comment]
There seems to be still an issue with this. I observe that both BASIS textures and legacy compressed textures break a cubemap used for skybox.

![image](https://user-images.githubusercontent.com/29543178/93865011-fd7ff600-fcc5-11ea-88bd-86ad3d0e0380.png)

Here is how it should look:

![image](https://user-images.githubusercontent.com/29543178/93864936-e3deae80-fcc5-11ea-9610-e40bfbaf7e89.png)

And here is how it looks after I compress the source files:

![image](https://user-images.githubusercontent.com/29543178/93864708-919d8d80-fcc5-11ea-9f28-d0b9a9e5fecb.png)

If I delete the prefiltered data, and try to prefilter again while the faces are compressed then It results in a black cubemap:

![image](https://user-images.githubusercontent.com/29543178/93865162-3cae4700-fcc6-11ea-8bef-3ddd34cb2277.png)

**Comment by slimbuck** [comment]
Hi @leonidaspir,

tldr: we flip textures upside-down during compression, this is why cubemaps are wrong.

This is related to the engine's treatment of texture origin. The PC engine follows the OpenGL convention of bottom left texture origin, whereas WebGL chose texture origin to be top left.

This means that we actually flip all (non-cubemap) textures upside-down:
- png/jpg textures are flipped at load time
- compressed textures are flipped at compression time
- all texture uvs are flipped to compensate

This system works most of the time, but fails for cubemaps (since these faces can't be flipped).

We are planning to address this by adopting the WebGL convention and not flipping assets. Not a small job though since we must also keep support for the existing flipped compressed textures.

In the meantime, you can get compressed cubemaps to work by using your own (non-flipped!) cubemap faces.

Thanks!
